File: javascript/node-postgres/src/alternatives/no_connection_pool/example_with_no_connection_pool.js
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

import assert from "node:assert";
import { AuroraDSQLClient } from "@aws/aurora-dsql-node-postgres-connector";

const ADMIN = "admin";
const NON_ADMIN_SCHEMA = "myschema";

async function getConnection(clusterEndpoint, user) {
  const client = new AuroraDSQLClient({
    host: clusterEndpoint,
    user: user,
    retry: { maxRetries: 5 },
  });

  await client.connect();
  return client;
}

async function example() {
  const clusterEndpoint = process.env.CLUSTER_ENDPOINT;
  assert(clusterEndpoint, "CLUSTER_ENDPOINT environment variable is not set");
  const user = process.env.CLUSTER_USER;
  assert(user, "CLUSTER_USER environment variable is not set");

  let client;
  try {
    client = await getConnection(clusterEndpoint, user);

    if (user !== ADMIN) {
      await client.query("SET search_path=" + NON_ADMIN_SCHEMA);
    }

    // Create a new table
    await client.query(`CREATE TABLE IF NOT EXISTS owner (
      id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
      name VARCHAR(30) NOT NULL,
      city VARCHAR(80) NOT NULL,
      telephone VARCHAR(20)
    )`);

    // Transactional write with OCC retry
    await client.transaction(async (c) => {
      await c.query(
        "INSERT INTO owner(name, city, telephone) VALUES($1, $2, $3)",
        ["John Doe", "Anytown", "555-555-1900"],
      );
    });

    // Check that data is inserted by reading it back
    const result = await client.query(
      "SELECT id, city FROM owner where name='John Doe'",
    );
    assert.deepEqual(result.rows[0].city, "Anytown");
    assert.notEqual(result.rows[0].id, null);

    await client.query("DELETE FROM owner where name='John Doe'");
    console.log("Completed successfully");
  } catch (error) {
    console.error(error);
    throw error;
  } finally {
    await client?.end();
  }
}

export { example };

File: java/pgjdbc/src/main/java/software/amazon/dsql/examples/alternatives/no_connection_pool/ExampleWithNoConnectionPool.java
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

package software.amazon.dsql.examples.alternatives.no_connection_pool;


import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.Properties;

import software.amazon.dsql.jdbc.OCCRetry;
import software.amazon.dsql.jdbc.OCCRetryConfig;

public class ExampleWithNoConnectionPool {

    // Get a connection to Aurora DSQL.
    public static Connection getConnection(String endpoint, String user) throws SQLException {
        Properties props = new Properties();
        props.setProperty("user", user);

        // Note: SSL is configured automatically by the connector with secure defaults.
        // No explicit SSL configuration needed.

        String url = "jdbc:aws-dsql:postgresql://" + endpoint;

        return DriverManager.getConnection(url, props);
    }

    public static void main(String[] args) throws SQLException {
        String clusterEndpoint = System.getenv("CLUSTER_ENDPOINT");
        assert clusterEndpoint != null : "CLUSTER_ENDPOINT environment variable is not set";

        String clusterUser = System.getenv("CLUSTER_USER");
        assert clusterUser != null : "CLUSTER_USER environment variable is not set";

        try (Connection conn = ExampleWithNoConnectionPool.getConnection(clusterEndpoint, clusterUser)) {
            if (!clusterUser.equals("admin")) {
                try (Statement setSchema = conn.createStatement()) {
                    setSchema.execute("SET search_path=myschema");
                }
            }

            // Use OCC retry with existing connection for write operations
            OCCRetryConfig retryConfig = OCCRetryConfig.defaults();

            // Create a new table named owner
            OCCRetry.execute(conn, retryConfig, c -> {
                try (Statement create = c.createStatement()) {
                    create.executeUpdate("""
                            CREATE TABLE IF NOT EXISTS owner(
                            id uuid NOT NULL DEFAULT gen_random_uuid(),
                            name varchar(30) NOT NULL,
                            city varchar(80) NOT NULL,
                            telephone varchar(20) DEFAULT NULL,
                            PRIMARY KEY (id))""");
                }
                return null;
            });

            // Insert some data
            OCCRetry.execute(conn, retryConfig, c -> {
                try (PreparedStatement insert = c.prepareStatement(
                        "INSERT INTO owner (name, city, telephone) VALUES (?, ?, ?)")) {
                    insert.setString(1, "John Doe");
                    insert.setString(2, "Anytown");
                    insert.setString(3, "555-555-1999");
                    insert.executeUpdate();
                }
                return null;
            });

            // Read back the data and assert they are present
            try (PreparedStatement read = conn.prepareStatement("SELECT * FROM owner WHERE name = ?")) {
                read.setString(1, "John Doe");
                try (ResultSet rs = read.executeQuery()) {
                    while (rs.next()) {
                        assert rs.getString("id") != null;
                        assert rs.getString("name").equals("John Doe");
                        assert rs.getString("city").equals("Anytown");
                        assert rs.getString("telephone").equals("555-555-1999");
                    }
                }
            }

            // Delete some data
            OCCRetry.execute(conn, retryConfig, c -> {
                try (PreparedStatement delete = c.prepareStatement("DELETE FROM owner WHERE name = ?")) {
                    delete.setString(1, "John Doe");
                    delete.executeUpdate();
                }
                return null;
            });
        }
        System.out.println("Connection exercised successfully");
    }
}
